### ISSUE TITLE
GRPO with tool use

### ISSUE THREAD

**Issue by catalin-hanga** [huggingface/trl#5366]
### Reproduction

I am trying to fine-tune the unsloth/Qwen3-0.6B model with GRPO and using a python REPL tool, analogus to this example:
https://huggingface.co/docs/trl/en/grpo_trainer#agent-training

I have configured the GRPO trainer in the following way:

```python
from trl import GRPOConfig, GRPOTrainer
trainer = GRPOTrainer(
    model = model,
    processing_class = tokenizer,
    reward_funcs = [reward_function],
    args =  GRPOConfig(...),
    train_dataset = dataset,
    tools = [run_python_code],
)
```
and the python REPL tool is defined like this:

```python
def run_python_code(code: str) -> str:
    """
    Executes Python code in a safe subprocess REPL.

    Args:
        code: python code to be executed.

    Returns:
        str: The output of the code.
    """
    
    import subprocess

    try:
        result = subprocess.run(
            ["python3", "-c", code],
            capture_output = True,
            check = False,
            text = True,
            timeout = 10,
        )
        return result.stdout or result.stderr
    except Exception as e:
        return str(e)
```

However, when I look at the completions generated by the LLM during training, I notice that none of them include any tool use.
Am I missing something from my code ? (the official online documentation explicitly says that the Qwen3-0.6B model has been tested for tool use with GRPO)

### System Info

Python version 3.12.12
device: one L4 GPU
OS: Ubuntu

vllm == 0.18.0
trl == 0.24.0
transformers == 4.56.2
unsloth == 2026.3.4
torch == 2.10.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by qgallouedec** [comment]
Your transformers version is too old, you need to update too at least 5.0. It will say that it's not compatible with vLLM, but you can safely force the install

**Comment by qgallouedec** [comment]
Btw, i recommend avoid the try/expect in the tool. The trainer will catch an error if the tool call fails

**Comment by catalin-hanga** [comment]
Thank you for the advice. I have updated my transformers library to version 5.2.0, and added an explicit instruction in the system prompt to use the tool, but it seems that the model is still unable to call it during training. (In fact, in one of the completions, I noticed the model saying: "But since the problem says to use the run_python_code tool, I need to actually perform this calculation. However, since I can't run code here, I'll have to rely on my approximations.")

If I print the full completions from inside the reward function, I should be able to see here that the model is calling the tool, at least in some of the runs, right ?
Also, does the GRPOTrainer automatically handle calling the tool, and then giving the tool result back to the LLM ?

**Comment by qgallouedec** [comment]
Hi @catalin-hanga,

I can't reproduce this. GRPO tool calling works fine. Here's a minimal repro (Qwen3-0.6B, one `multiply` tool, a trivial "A times B?" task), runs on a single GPU with no vLLM:

```python
import random

from datasets import Dataset

from trl import GRPOConfig, GRPOTrainer


def multiply(a: int, b: int) -> int:
    """
    Multiplies two integers.

    Args:
        a: The first integer.
        b: The second integer.

    Returns:
        The product of the two integers.
    """
    return a * b


SYSTEM = (
    "You are a careful calculator. To answer, you MUST call the `multiply` tool. "
    "After getting the tool result, give the final answer."
)

rng = random.Random(0)
rows = []
for _ in range(64):
    a, b = rng.randint(11, 99), rng.randint(11, 99)
    rows.append({
        "prompt": [{"role": "system", "content": SYSTEM}, {"role": "user", "content": f"What is {a} times {b}?"}],
        "answer": a * b,
    })
dataset = Dataset.from_list(rows)


def reward_correct(completions, answer, **kwargs):
    return [float(str(ans) in " ".join(m.get("content") or "" for m in c)) for c, ans in zip(completions, answer)]


trainer = GRPOTrainer(
    model="Qwen/Qwen3-0.6B",
    reward_funcs=[reward_correct],
    args=GRPOConfig(
        output_dir="trainer_output/mre_5366",
        per_device_train_batch_size=8,
        num_generations=8,
        max_completion_length=256,
        max_steps=5,
        logging_steps=1,
        bf16=True,
        log_completions=True,
        use_vllm=False,
    ),
    train_dataset=dataset,
    tools=[multiply],
)
trainer.train()
```

A sample completion shows the full tool round-trip, and the per-step metrics confirm tools are called throughout:

```
<tool_call>
{"name": "multiply", "arguments": {"a": 88, "b": 29}}
</tool_call>
<tool_response>
2552
</tool_response>

step 1: tools/call_frequency: 0.75,  tools/failure_frequency: 0, reward: 0.875
step 2: tools/call_frequency: 0.75,  tools/failure_frequency: 0, reward: 1.0
step 3: tools/call_frequency: 0.875, tools/failure_frequency: 0, reward: 1.0
step 4: tools/call_frequency: 0.75,  tools/failure_frequency: 0, reward: 1.0
step 5: tools/call_frequency: 0.875, tools/failure_frequency: 0, reward: 1.0
```

So the mechanism works. Your two questions: **yes**, the completions passed to the reward function include the assistant `tool_calls` and `tool` response messages; and **yes**, `GRPOTrainer` executes the tool and feeds the result back automatically.

What was likely wrong in your setup:
1. **transformers version**: tools need `>=5.0`; make sure the env you run actually uses 5.2.0.
2. **Prompt the model to use the tool**: a 0.6B model won't call a tool unless told to (note the explicit system prompt). Your "I can't run code here" completion suggests the model wasn't even aware a tool was available.
3. **unsloth tokenizer/template**: load the stock model with `model="Qwen/Qwen3-0.6B"`; unsloth's patched template may not render tools or be prefix-preserving, which TRL relies on. I have no idea tbh, but it's a possibility

I'll close this since it's not a bug, but feel free to reopen if the MRE above doesn't work for you.
